Implementation sensor device (#126)
SW709E19/Report: report/ChapterImplementation/SensorDevice.tex

Among the devices the that came onto consideration were the Arduino Uno with a Wi-Fi shield in order to communicate wirelessly with the gateway as described in~\autoref{subsec:designArchitectureDiagram}.
This device would fit within the design criteria and was thus a possible candidate\cite{ArduinoCompare}.

Another device was the ESP32 that comes with both Wi-Fi and Bluetooth builtin\cite{ESP32}.

Between the two there are many similarities, but the ESP32 are able to go into a state called deep sleep were its power consumption is a $10\mu A$ \cite{ESP32}.
This property makes the ESP32 (Show on~\autoref{fig:esp32}) type of devices a more suitable choice for the project since the device should be powered by a battery.
The overall construction and since of the device makes it a more suitable choice to incorporate into a production environment at a factory due to is small form factor.

\figur{0.5}{ChapterImplementation/images/ESP32S.jpg}{ESP32 Development board \cite{ESP32ArduinoTech}.}{fig:esp32}

The development board was chosen do to its ease of use for non technical people and due to its low price of 79,00 dkk\cite{ESP32ArduinoTech}. 

> The supplier of the device was expected to be with in a few days, since the seller located in Denmark.

Reviewer: ```suggestion
The device was expected to be delivered within a few days, since the seller is located in Denmark.
```

> Based on recommendations from Rasmus Steiniche the most important sensor that the system should support was an accelerometer.

Reviewer: Vil det give mening at vi undlader sætninger, hvor vi siger noget om hvad Rasmus har sagt uden at citere ham?
```suggestion
In collaboration with the product owner it was decided that the first supported sensor in the system should be an accelerometer.
```

However, the supplier of the ESP32s had a very limited supply of accelerometers and thus those could not be ordered from the same supplier.
Luckily one of the group members had access to an accelerometer suitable for the project.

> The accelerometer is an MPU6050 which contains both an accelerometer, a gyroscope and a temperature sensor with in once chip.

Reviewer: ```suggestion
The accelerometer is an MPU6050 which contains both an accelerometer, a gyroscope and a temperature sensor within the same chip.
```

%IGNORE_LINE
The sensor can communicate with the ESP32 via the $I^2C$ protocol via the analogue pins \cite{MPUdatasheet}.
The sensor can be attached to the ESP32 as seen on~\autoref{fig:SensorDeviceAssembly} and does not require a lot of technical skill to assemble.

\figur{1}{ChapterImplementation/images/ESP32Side.jpg}{ESP32 with a accelerometer attached.}{fig:SensorDeviceAssembly}

\subsection{Software}

% Based on the selected hardware tell how it can be programed.

% Tell about the different configurations.

% Tell about MQTT and WIFI configurations on the esp32.

% Tell about the configuration if the accelerometer. 

% Show some important/cool/hyped code, if any exists. 

With the hardware in place the development of the sensor device could start.

> Since the ESP32 devices are very similar to the Arduino type they can be programmed using the same development tools and they also support the same programming languages.

Reviewer: ```suggestion
```

> The code that runs on the sensor device can be found on the \href{https://github.com/SW709E19/Product/tree/develop/sensor-device}{GitHub repository} and is publicly available.

Reviewer: Jeg synes vi skal samle henvisninger til GitHub repo til ét sted i rapporten. Synes ikke det er vigtigt her.
```suggestion
```

As described in~\autoref{subsec:designSensorDevice} the sensors should support wireless communication between it self and the gateway.
This was achieved using the onboard Wi-Fi module on the ESP32.

> cat lin	

Reviewer: ```suggestion
```

%IGNORE_LINE
This was easily achieved using the "WiFi.h" header file for the ESP32.
Once included a connection could be established with a method call with the Service Set Identifier (SSID) and password as parameters.

Once connected to a gateway the device should start collecting measurements and since the sensor is connected via the analogue pins on the ESP32 some configuration is needed to communicate.

> The sensor module has an integrated controller that stores the measurements before they are sent to the ESP32 via the $I^2C$ protocol.

Reviewer: ```suggestion
During each duty cycle the ESP32 reads measurements from the MPU6050 via the $I^2C$ protocol.
```

> One the data is received by the ESP32 it should be sent to the gateway for further processing.

Reviewer: ```suggestion
Once the data is received by the ESP32 it should be sent to the gateway for further processing.
```

To achieve this the MQ Telemetry Transport (MQTT) protocol was used.
The protocol was chosen with respect to the design criteria described in~\autoref{cha:communication} and from the experiences gathered during the prototype development.

> MQTT is a lightweight machine-to-machine connectivity protocol and was design as a extremely lightweight publish/subscribe messaging transport \cite{MQTT}.

Reviewer: Minder meget om det der står direkte på mqtt.org. Jeg tænker at vi bare kan citere direkte i stedet for at forsøge at omskrive deres beskrivelse..
```suggestion
\textit{``MQTT is a machine-to-machine (M2M)/"Internet of Things" connectivity protocol. It was designed as an extremely lightweight publish/subscribe messaging transport.''} \cite{MQTT}.
```

The ESP32 has a client library that enables support for the MQTT protocol which makes it easy to implement.
Once the ESP32 is connected to the Wi-Fi it will establish a connection to the MQTT server running on the gateway.
Once the device returns from deep sleep it will subscribe to the config channel at the MQTT server.

> Here it will publish it current configuration to the state topic and receive a new configuration of it is no longer valid.

Reviewer: ```suggestion
Here it will publish its current configuration to the state topic and receive a new configuration if it is no longer valid.
```

> Each configuration currently contains four different properties; sample duration, sleep duration and sample rate

Reviewer: ```suggestion
Each configuration currently contains three different properties; sample duration, sleep duration and sample rate.
```

> The sample duration specifies how for how long the sensor devices should take samples.

Reviewer: ```suggestion
The sample duration specifies for how long the sensor devices should read measurements.
```

The sleep duration specifies for how many seconds the sensor devices should be in deep sleep mode.

> The sample rate specifies how fast the sensor should sample.

Reviewer: ```suggestion
The sample rate specifies how many measurements should be collected per second.
```

> This leaves it up to the system architect to fine tune the configuration to each specific devices.

Reviewer: ```suggestion
This leaves it up to the system architect to fine tune the configuration to each specific device.
```

Throughout this section a deeper understanding of the implementation will be established.
The different parts of the sensor device will be elaborated upon through the implementation of the system.

\subsection{Hardware}

% Tell about the hardware we got and why we got it(time, price, easy to use for unexperienced users)

% Describe some of the different considerations made and relate them to "Design - Architecure - Sensor Devices.

% Show the hardware, with a picture of individual components and describe each of them.

% Show how the are connected, with a picture.

Before an implementation could start, the most basic part was needed ie. the hardware.

Based on the description in~\autoref{subsec:designSensorDevice} the search for a suitable device was conducted.
With respect to the limited time for this project, the ideal solution would be to obtain some hardware that the university or someone within the project group already had.
However since the group retains under the institute of computer science the institute had a very limited amount of hardware to take into consideration.

> Thus an application for project funding was made to the ICT-committee at the university based on the application procedure that can be seen in~\autoref{app:ICTApplicationProcedure}. \todo{Perhaps not allowed}

Reviewer: Jeg synes ikke det er relevant for rapporten hvordan vi har fået hardwaren. Det vigtigste er at det er tydeligt hvorfor esp32 er et godt valg i forhold til andre MCU'er. Evt. med hensyn til de tre: cheap, flexible, easy to install

In the application the hardware wanted needed to be specified in order for it to be approved by the committee.

Based on earlier experiences and research online suitable devices was found.

> Among the devices the that came onto consideration were the Arduino with a Wi-Fi shield in order to communicate wirelessly with the gateway as described in~\autoref{subsec:designArchitectureDiagram}.

Reviewer: Hvilken Arduino? Uno, Mega, Nano?

This device would fit within the design criteria and was thus a possible candidate\cite{ArduinoCompare}.

> Another device was the ESP32 type system that comes with both Wi-Fi and Bluetooth builtin\cite{ESP32}.

Reviewer: type system?

> Between the two there are many similarities, but the ESP32 type devices are able to go into a state called deep sleep were its power consumption is a a minimum.

Reviewer: I stedet for power consumption is a a minimum., så kunne du tilføje deep sleep current for datasheetet for esp32

This property makes the ESP32 (Show on~\autoref{fig:esp32}) type of devices a more suitable choice for the project since the device should be powered by a battery.

\figur{0.5}{ChapterImplementation/images/ESP32S.jpg}{ESP32 Development board \cite{ESP32ArduinoTech}.}{fig:esp32}

The development board was chosen do to its ease of use for non technical people and due to its low price of 79,00 dkk\cite{ESP32ArduinoTech}. 
The supplier of the device was expected to be with in a few days, since the seller located in Denmark.

All these considerations all resulted in an application for the ICT committee for 4 ESP32 and some other hardware such as cables and USB power tester.

> The application was approved with funding for $\sim$ 800,00 dkk.

Reviewer: Jeg synes ikke det er relevant at nævne ansøgning og funding her.

Once included a connection could be established with a method call with the Service Set Identifier (SSID) and password as parameters.

Once connected to a gateway the device should start collecting measurements and since the sensor is connected via the analogue pins on the ESP32 some configuration is needed to communicate.
The sensor module has an integrated controller that stores the measurements before they are sent to the ESP32 via the $I^2C$ protocol.

One the data is received by the ESP32 it should be sent to the gateway for further processing.
To achieve this the MQ Telemetry Transport (MQTT) protocol was used.
The protocol was chosen with respect to the design criteria described in~\autoref{cha:communication} and from the experiences gathered during the prototype development.
MQTT is a lightweight machine-to-machine connectivity protocol and was design as a extremely lightweight publish/subscribe messaging transport \cite{MQTT}.

The ESP32 has a client library that enables support for the MQTT protocol which makes it easy to implement.
Once the ESP32 is connected to the Wi-Fi it will establish a connection to the MQTT server running on the gateway.

> Once connected it will subscribe on the config topic and download the latest config set to the device.

Reviewer: Sensor device gør følgende når den vågner fra deep sleep:
- subscriber til configuration topic
- publisher sin nuværende configuration til state topic
- modtager ny configuration, hvis den sendte ikke længere er gældende

The sensor device runs with different user defined configurations and they can be changed "on the fly" by the system architect and will be updated every time the devices connects to the gateway.

> Each configuration currently contains four different properties; sample duration, sleep duration, sample rate and bulk size.

Reviewer: Bulk size er fjernet (Den gav ikke så meget mening)
